drivers/spi/spi_wch.c

```diff
@@ -0,0 +1,259 @@
+	if (err != 0) {
+		return err;
+	}
+
+	/* Approximate clock rate given ratios available */
+	div = spi_wch_get_br(clock_rate / config->frequency);
+#if CONFIG_SPI_LOG_LEVEL >= LOG_LEVEL_INF
+	uint32_t j = 2;
+	for (int i = 0; i < div; i++) {
+		j = j * 2;
+	}
+	LOG_INF("Selected divider %d, value %d, results in %d frequency", j, div, clock_rate / j);
+#endif
+	cfg->regs->CTLR1 |= (uint8_t)div << SPI_CTLR1_BR_POS;
+
+	data->ctx.config = config;
+
+	return 0;
+}
+
+static int spi_wch_transceive(const struct device *dev,
+			       const struct spi_config *config,
+			       const struct spi_buf_set *tx_bufs,
+			       const struct spi_buf_set *rx_bufs)
+{
+	const struct spi_wch_config *cfg = dev->config;
+	struct spi_wch_data *data = dev->data;
+	int err;
+	uint8_t rx;
+
+	spi_context_lock(&data->ctx, false, NULL, NULL, config);
+
+	err = spi_wch_configure(dev, config);
+	if (err != 0) {
+		goto done;
+	}
+
+	spi_context_buffers_setup(&data->ctx, tx_bufs, rx_bufs, 1);
+
+	spi_context_cs_control(&data->ctx, true);
+
+	/* Start SPI *AFTER* setting CS */
+	cfg->regs->CTLR1 |= SPI_CTLR1_SPE;
+
+	while (spi_context_tx_on(&data->ctx) || spi_context_rx_on(&data->ctx)) {
+		if (spi_context_tx_buf_on(&data->ctx)) {
+			cfg->regs->DATAR = *(uint8_t *)(data->ctx.tx_buf);
+			while(!(cfg->regs->STATR & SPI_STATR_TXE)) {}
+		} else {
+			cfg->regs->DATAR = 0;
+		}
+		spi_context_update_tx(&data->ctx, 1, 1);
+		while(!(cfg->regs->STATR & SPI_STATR_RXNE)) {}
+		rx = cfg->regs->DATAR;
+		if (spi_context_rx_buf_on(&data->ctx)) {
+			*data->ctx.rx_buf = rx;
+		}
+		spi_context_update_rx(&data->ctx, 1, 1);
+	}
+
+	spi_context_cs_control(&data->ctx, false);
+
+done:
+	cfg->regs->CTLR1 &= ~(SPI_CTLR1_SPE);
+	spi_context_release(&data->ctx, err);
+	return err;
+}
+
+static int spi_wch_transceive_sync(const struct device *dev,
+				    const struct spi_config *config,
+				    const struct spi_buf_set *tx_bufs,
+				    const struct spi_buf_set *rx_bufs)
+{
+	return spi_wch_transceive(dev, config, tx_bufs, rx_bufs);
+}
+
+static int spi_wch_release(const struct device *dev,
+			    const struct spi_config *config)
+{
+	struct spi_wch_data *data = dev->data;
+
+	spi_context_unlock_unconditionally(&data->ctx);
+
+	return 0;
+}
+
+static int spi_wch_init(const struct device *dev)
+{
+	int err;
+	const struct spi_wch_config *cfg = dev->config;
+	struct spi_wch_data *data = dev->data;
+	clock_control_subsys_t clk_sys;
+
+	clk_sys = (clock_control_subsys_t)(uintptr_t)cfg->clk_id;
+
+	err = clock_control_on(cfg->clk_dev, clk_sys);
+	if (err < 0) {
+		return err;
+	}
+
+	err = pinctrl_apply_state(cfg->pcfg, PINCTRL_STATE_DEFAULT);
+	if (err < 0) {
+		return err;
+	}
+
+	err = spi_context_cs_configure_all(&data->ctx);
+	if (err < 0) {
+		return err;
+	}
+
+	spi_context_unlock_unconditionally(&data->ctx);
+
+	return 0;
+}
+
+static DEVICE_API(spi, spi_wch_driver_api) = {
+	.transceive = spi_wch_transceive_sync,
+#ifdef CONFIG_SPI_RTIO
```

Comment:
Should this be here?

Reply:
yes, it's to handle RTIO interactions with a driver that doesnt support RTIO. Not sure if it requires async in the first place though.

Reply:
Ah, I assumed it was copy-paste from a driver called "spi_rtio".